lib/saluki-components/src/destinations/datadog/common/io.rs

```diff
@@ -193,33 +210,48 @@ async fn run_endpoint_io_loop<S, B>(
 
     // Build our endpoint service.
     //
-    // This is where we'll modify the incoming request for our our specific endpoint, such as setting the host portion
+    // This is where we'll modify the incoming transaction for our our specific endpoint, such as setting the host portion
     // of the URI, adding the API key as a header, and so on.
     let mut service = ServiceBuilder::new()
         // Set the request's URI to the endpoint's URI, and add the API key as a header.
-        .map_request(for_resolved_endpoint::<B>(endpoint))
+        .map_request(for_resolved_endpoint(endpoint))
         // Set the User-Agent and DD-Agent-Version headers indicating the version of the data plane sending the request.
-        .map_request(with_version_info::<B>())
+        .map_request(with_version_info())
         .concurrency_limit(config.endpoint_concurrency())
         .service(service);
 
+    let retry_queue = RetryQueue::new(endpoint_url.clone(), config.retry().retry_queue_max_size_bytes);
+    let mut pending_transactions = PendingTransactions::new(config.endpoint_buffer_size(), retry_queue);
+
     let mut in_flight = JoinSet::new();
     let mut done = false;
 
     loop {
         select! {
```

Comment:
Rust question: `select` just chooses the branch that completes first right? Is there a way to prioritize which one runs? (Let's say maybe we want to finish all requests to completion before starting new requests.

Reply:
Yes, and yes.

By default, the `select!` will generate code to randomize the order it checks the branches in. You can add a little extra directive -- `biased;` -- which turns that off and always checks them in the order they're defined in:

```rust
select! {
  biased;

  _ = my_future => { ... }
}
```

Generally you want the randomization for fairness, and to potentially avoid deadlocking yourself if you're doing something complex... but there are situations where biased polling makes sense, like if one of your branches is checking if you should shutdown, since otherwise you might take another branch (perhaps one that causes you to do a lot of work or go to sleep) a few times before you take the shutdown one, and so on.